larky/src/main/resources/stdlib/urllib/request.star

```diff
@@ -1,268 +1,294 @@
+            # if we change data we need to remove content-length header
+            # (cause it's most probably calculated for previous value)
+            if self.has_header("Content-length"):
+                self.remove_header("Content-length")
+    self.data = larky.property(_get_data, _set_data)
+    self.body = larky.property(_get_data, _set_data)
+
+    def _parse():
+        self.type, rest = parse._splittype(self._full_url)
+        if self.type == None:
+            fail("ValueError: " + "unknown url type: %r" % self.full_url)
+        self.host, self.selector = parse._splithost(rest)
+        if self.host:
+            self.host = parse.unquote(self.host)
+    self._parse = _parse
+
+    def get_method():
+        """Return a string indicating the HTTP request method."""
+        default_method = "POST" if self.data != None else "GET"
+        return getattr(self, "_method", default_method)
+    self.get_method = get_method
+    self.method = larky.property(get_method)
+
+    def get_full_url():
+        return self.full_url
+    self.get_full_url = get_full_url
+
+    def set_proxy(host, type):
+        if self.type == "https" and not self._tunnel_host:
+            self._tunnel_host = self.host
+        else:
+            self.type = type
+            self.selector = self.full_url
+        self.host = host
+    self.set_proxy = set_proxy
+
+    def has_proxy():
+        return self.selector == self.full_url
+    self.has_proxy = has_proxy
+
+    def add_header(key, val):
+        # original implementation uses key.capitalize(), however, we will not modify the keys.
+        self._headers[key] = val
+    self.add_header = add_header
+
+    def add_unredirected_header(key, val):
+        # will not be added to a redirected request
+        # original implementation uses key.capitalize(), however, we will not modify the keys.
+        self.unredirected_hdrs[key] = val
+    self.add_unredirected_header = add_unredirected_header
+
+    def has_header(header_name):
+        return header_name in self.headers or header_name in self.unredirected_hdrs
+    self.has_header = has_header
+
+    def get_header(header_name, default=None):
+        return self.headers.get(
+            header_name,
+            self.unredirected_hdrs.get(header_name, default)
+        )
+    self.get_header = get_header
+
+    def remove_header(header_name):
+        self.headers.pop(header_name, None)
+        self.unredirected_hdrs.pop(header_name, None)
+    self.remove_header = remove_header
+
+    def header_items():
+        hdrs = {}
+        hdrs.update(self.unredirected_hdrs)
+        hdrs.update(self._headers)
+        return list(hdrs.items())
+    self.header_items = header_items
+
+    def _get_headers():
+        return self._headers
+
+    def _set_headers(headers):
+        self._headers = {}
+        for key, value in headers.items():
+            self.add_header(key, value)
+    self.headers = larky.property(_get_headers, _set_headers)
```

Comment:
it doesn't look like a property to me: https://github.com/python/cpython/blob/3.10/Lib/urllib/request.py#L323